plugin/core/windows.py

```diff
@@ -557,6 +566,48 @@ def _check_window_closed(self) -> None:
         if not self._is_closing and not self._window.is_valid():
             self._handle_window_closed()
 
+    def _receive_progress_token(self, params: Dict[str, Any], client: Client, request_id: Any) -> None:
+        token = params.get('token')  # number | string
```

Comment:
For mandatory keys in the spec, I would use `params['token']`. This frees up a lot of clutter and error checking.
If a `KeyError` is raised because of this, then it is the language server's fault. Any exception in any notification handler is caught and handled, so the plugin won't break because of faulty language servers.

For instance, below would become simply `self._progress[params['token']] = dict()`.